csa/man/csaaddc.8.in
.\"
.\" (C) 2000-2007 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This documentation supports a
.\" program that is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\"
.\" The documentation that supports
.\" this program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" Further, this documentation and the software it
.\" supports, are distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\"
.\" Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
.\" Sunnyvale, CA  94085, or:
.\"
.\" http://www.sgi.com
.\"
.\" For further information regarding this notice, see:
.\"
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan
.\"
.TH csaaddc 8
.SH NAME
\f4csaaddc\f1 \- combines \f4cacct\f1 records
.SH SYNOPSIS
\f4@sbindir@csaaddc\f1
[\f4-A\f1]
[\f4-D\f1\ \f2level\f1]
[\f4-o\f1\ \f2ofile\f1]
[\f4-t\f1]
[\f4-v\f1]
[[[\f4-g\f1]
[\f4-p\f1]]
[\f4-j\f1]
[\f4-u\f1]]
\f2input-files\f1
.SH DESCRIPTION
The \f4csaaddc\f1 command combines \f7cacct\f1 records
by specified consolidation options and writes out a
consolidated record in either \f7cacct\f1 or ASCII
format.
You also can generate \f7cacct\f1 records with the
\f4csacon\f1(8), \f4acctdisk\f1(8) and \f4csachargefee\f1(8) commands.
.PP
The \f4csaaddc\f1 command accepts the following operand:
.TP 10
\f2input-files\f1
Specifies the files to be processed.
The files are in \f7cacct\f1 format.
.SS Output Options
.TP 10
\f4-A\f1
Outputs information in ASCII.
By default, the output is written to a binary output file.
.TP
[\f4-D\f1\ \f2level\f1]
Sets debugging level.
Level 1 is somewhat verbose; level 2 is verbose.
.TP
\f4-o\f1\ \f2ofile\f1
Specifies the output file.
By default, the output is written to \f7stdout\f1.
.TP
\f4-t\f1
Consolidates all records into one total record.
.TP
\f4-v\f1
Sets verbose mode on.
When you also specify \f4-A\f1,
verbose output is written to \f7stderr\f1.
.SS Consolidation Options
You can specify multiple consolidation options.
These options should be the same as those used
to create the input files with \f4csacon\f1(8).
If you do not specify any consolidation options,
the default is \f4-pu\f1.
.TP 10
\f4-g\f1
Consolidates records by group ID.
You must use this option with at least one of the other
consolidation options, because not all records have a group ID.
.TP
\f4-p\f1
Consolidates records by project ID. (deferred)
.TP
\f4-j\f1
Consolidates records by job ID.
.TP
\f4-u\f1
Consolidates records by user ID.
.SH NOTES
The consolidation options used with \f4csacon\f1(8)
to generate the input files should correspond to the
consolidation options used with \f4csaaddc\f1.
If they do not, the resulting data can be
misleading and difficult to interpret.
.SH EXAMPLES
The following example merges two input files
created with \f4csacon\f1(8), using the
\f4-g\f1 consolidation option.
The output is written to file \f7outfile\f1.
.RS .5i
.PP
\f4csaaddc  -g -o outfile cacct1 cacct2\f1
.RE
.SH SEE ALSO
.PP
acctdisk(8),
csachargefee(8),
csacon(8),
csacrep(8),
dodisk(8).


csa/ChangeLog
* Wed Jun  8 2016 Jay Lan <jay.j.lan@nasa.gov>
- Converted csa package from the last SGI-supported version 4.1.1 to
  community version 5.0.0

* Fri Oct 17 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.1
- Various fixes to allow for SLES11 build

* Fri Jul 25 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.0
- Fixed memory leaks
- Changed acctjob structure offset type from int to int64_t

* Fri Apr 18 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.2
- Updated to use new libnl API
- Fixed process exit code assignment

* Tue Mar 4 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.1
- Cleaned up acctdusg
- Added check to prevent SIGPIPE when communicating with daemon
- Increased maximum PID allowed for larger systems

* Wed Nov 7 2007 Jonathan Lim <jlim@sgi.com>
- csa-4.0.0
- Implement kernel functionality as a userspace daemon
- Remove obsolete wait metrics and replace with taskstats delay accounting
- Various bug fixes

* Fri Apr 27 2007 Jay Lan <jlan@sgi.com>
* Mon Apr 30 2007 Jonathan Lim <jlim@sgi.com>
- csa-3.0.2
- SBU's data was all zero in csa-3.0.0 and csa-3.0.1
- Clarify csacom(1) man page on '-f' option
- Obsolete csacom '-a' & '-q' option
- Fix a problem of csadrep reading input from a binary file
- Fixed csadrep(8) man page: replaced '-w' with '-n' and removed '-t'

* Wed Jul 5 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.1
- 'csacom -b' was broken in csa-3.0.0
- 'csacom -b' drops the last record of a pacct file

* Thu Feb 24 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.0
- csacom -[esES] option was broken.
- create /usr/sbin/rccsa symlink at %install in the spec file
- remove code for CSA record types not supported in Linux
- spliting up csa_kern.h into kernel version and userland version
- support pacct files created in csa-2.*.*

* Wed Jul 27 2005 Chris Sturtivant <csturtiv@sgi.com>
- csa-2.2.2
- Disabled perror for permission denied in csa_ctl when called by csa_auth

* Mon May 2 2005 Jay Lan <jlan@sgi.com>
- csa-2.2.1
- clean up csa init script
- csa_wracct(3) man page update
- 'csacom -o' wrote only one record
- 'csacom -b' does not work
- csajrep man page references invalid option '-W'
- 'csajrep -r' does not work
- 'dodisk' created files with incorrect user and group name 'adm'
- CSA does not handle gid's > 65535
- 'csacom -j' failed if job id greater than max value of an int

* Mon Nov 22 2004 Jay Lan <jlan@sgi.com>
- csa-2.2.0
- Fixed compilation issue with gcc 3.3.3
- CSA assumes existence of 'adm' user/group account
- csacom & csacms usage otuput missing carriage return
- csarecy misses a carriage return on certain condition
- csacom -W segfaults after printing one bad time

* Fri Oct 8 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.2
- Fixed 'csaswitch -c halt' bug
- Fixed 'csaedit dumped core on bad wkmg record' bug

* Tue Jul 27 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.1
- Remove static API library libcsa.a from rpm package
- Fixed a bug at csacom with invalid -u option
- cascom headers and data not aligned correctly at output

* Wed Jun 30 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.0
- Fixed a problem dealing with accounting files > 2GB
- Now derives "Day of Year" from other fields of the holidays file
- Removed "Pages Swapped" from csacom and csacms - not supported in Linux
- Fixed a bug of not mapping UID 0 to root

* Tue Feb 17 2004 Jay Lan <jlan@sgi.com>
- csa-2.0.0
- Implemented /proc ioctl interface
- Renamed rpm from csa-job-acct to csa
- Renamed /etc/init.d/csaacct to /etc/init.d/csa

* Fri Nov 7 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-2.0.0
- Provided API library libcsa.so and associated header file

* Wed Sep 3 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-1.0.3
- Updated the spec file

* Thu Feb 13 2003 Robin Holt <holt@sgi.com>
- Updated init scripts to use /var/lock/subsys

* Tue Nov 5 2002 Richard Offer <offer@sgi.com>
- Fixed packager/distribution/vendor

* Thu May 23 2002 Robin Holt <holt@sgi.com>
- csa-job-acct-1.0.2
- Changed install destination from /usr/local to /usr
- Updated the man page location to /usr/share/man
- Added a clean and realclean target to the Makefile

* Thu Apr 5 2001 Marlys Kohnke <kohnke@sgi.com>
- Changed csadrep internal markers back to match those in IRIX CSA
- Changed ja summary report so that all lines are displayed even if the
  resource used values are zero

* Sun Feb 11 2001 Marlys Kohnke <kohnke@sgi.com>
- Added support for Workload Management daemon accounting records
- Fixed csachkpacct script problem (df -k needs to be df -kP; changed awk
  field to 4)

* Fri Jan 19 2001 Daryl Grunau <dwg@acl.lanl.gov>
- port to Linux from SGI IRIX
- csa-job-acct-1.0.0 (available January 19, 2001)
- initial csa-job-acct-1.0.0 open source release


csa/cmd/csaaddc.c
/*
 * Copyright (c) 2000-2002 Silicon Graphics, Inc and LANL  All Rights Reserved.
 * Copyright (c) 2004-2007 Silicon Graphics, Inc All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as 
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it would be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details. 
 *
 * You should have received a copy of the GNU General Public License along 
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston MA 02111-1307, USA.
 *
 * Contact information:  Silicon Graphics, Inc., 1140 East Arques Avenue,
 * Sunnyvale, CA  94085, or:
 * 
 * http://www.sgi.com 
 */

/*
 *	csaaddc.c - add cacct records together.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif	/* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>

#include <sys/param.h>
#include <sys/types.h>
#include "csaacct.h"

#include <string.h>
#include <time.h>
#include <unistd.h>

#include "acctdef.h"
#include "acctmsg.h"
#include "cacct.h"

extern int db_flag;

int	errflag		= 0;
char	*Prg_name		= "";
char	*outfile	= NULL;
FILE	*ofd;
int	bofd		= -1;
int	ifd		= -1;

struct	map	{
	uint64_t	prid;
	uid_t		uid;
	uint64_t	jid;
	gid_t		gid;
	struct	cacct	*acp;
};

struct	map	*acmap	= NULL;
#define	MAPSIZE		1000		/* initial map size */
#define	MAPINC		250		/* map increment */

int	mapsize	= 0;
int	total	= 0;
struct	cacct	tcb;			/* total structure if requested */
int	verbose	= 0;
int	asciiout = 0;
int	con_prid = 0;
int	con_gid = 0;
int	con_jid = 0;
int	con_uid = 0;

static	void	addmap(uid_t, uint64_t, uint64_t, gid_t, int);
static	void	enter();
static	struct	cacct	*findacd(uid_t, uint64_t, uint64_t, gid_t);
static	void	wrascii(struct cacct *);
static	void	wroutput();

static	void
usage() {
	acct_err(ACCT_ABORT,
	       "Command Usage:\n%s\t[-A] [-D debug] [-jpu [-g]] [-o ofile] [-t] [-v] ifile(s)",
		Prg_name);
}


main(int argc, char **argv)
{
	char		ch;
	int		c;
	extern char	*optarg;
	extern int	optind;

	Prg_name = argv[0];
	while((c = getopt(argc, argv, "AD:gjo:tuv")) != EOF) {
		ch = (char)c;
		switch(ch) {
		case 'A':
			asciiout = 1;
			break;
		case 'D':
			db_flag = atoi(optarg);
			if ((db_flag < 0) || (db_flag > 2)) {
				acct_err(ACCT_FATAL,
				       _("The (-%s) option's argument, '%s', is invalid."),
					"D", optarg);
				Nerror("Option -D valid values are 0 thru 2\n");
				usage();
			}
			break;
		case 'g':
			con_gid = 1;
			break;
		case 'j':
			con_jid = 1;
			break;
		case 'o':
			outfile = optarg;
			break;
		case 'p':
			con_prid = 1;
			break;
		case 't':
			total = 1;
			break;
		case 'u':
			con_uid = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			usage();
		} /* end switch */
	} /* end while getopt */

	/*
	 *	If no consolidation options were specified, use -pu.
	 */
	if ((con_prid + con_gid + con_jid + con_uid) == 0) {
		con_prid = con_uid = 1;
	}

	/*
	 *	If group only, usage error since not all records have gids.
	 */
	if ((con_prid + con_jid + con_uid) == 0) {
		usage();
	}

	/*
	 *	Open the output file
	 */
	if (asciiout) {
		bofd = -1;
		ofd = outfile == NULL ? stdout :  fopen (outfile,"w");
	} else {
		ofd = NULL;
		bofd = outfile == NULL ?
			openacct(NULL, "w") : openacct(outfile, "w");
	}
	if ((bofd < 0) &&
	    (ofd == NULL)) {
		acct_perr(ACCT_ABORT, errno,
			_("An error occurred during the opening of file '%s'."),
			outfile ?
			outfile : "stdout");
	}

	/*
	 *	Make sure we have at least 1 input file
	 */
	if (argv[optind] == NULL) {
		acct_err(ACCT_FATAL,
		       _("The input file was not specified.")
			);
		usage();
	}

	/*
	 *	Process each inputfile.
	 */
	for( ; optind < argc; optind++) {
		if ((ifd = openacct(argv[optind], "r")) < 0) {
			acct_perr(ACCT_ABORT, errno,
				_("An error occurred during the opening of file '%s'."),
				argv[optind]);
		}
		enter();
		closeacct(ifd);
	}

	/*
	 *	Write output file.
	 */
	if(total) {
		tcb.ca_prid = 0;
		tcb.ca_uid = 0;
	}
	wroutput();
	if (bofd > 0) {
		closeacct(bofd);
	}

	exit(errflag);
}

/*
 *	Routine to summmarize records...
 */
static void
enter() 
{
	struct cacct *acp;
	struct	cacct cb;
	int ind;
	struct	cacct	*findacd();

	while (readcacct (ifd, &cb) == sizeof (struct cacct)) {
		if (total) {
			acp = &tcb;
		} else {
			if (db_flag > 1) {
				fprintf(stderr, "enter(): calling findacd() "
				    "with uid %d (%s), prid %lld (%s), "
				    "jid 0x%llx.\n", cb.ca_uid,
				    uid_to_name(cb.ca_uid), cb.ca_prid,
				    prid_to_name(cb.ca_prid), cb.ca_jid);
			}
			if ((acp = findacd(cb.ca_uid, cb.ca_prid, cb.ca_jid,
				cb.ca_gid)) == (struct cacct *)NULL) {
				exit(-1);
			}
		}
		acp->ca_njobs += cb.ca_njobs;
		acp->prime.cpuTime += cb.prime.cpuTime;
		acp->nonprime.cpuTime += cb.nonprime.cpuTime;
		acp->prime.charsXfr += cb.prime.charsXfr;
		acp->nonprime.charsXfr += cb.nonprime.charsXfr;
		acp->prime.logicalIoReqs += cb.prime.logicalIoReqs;
		acp->nonprime.logicalIoReqs += cb.nonprime.logicalIoReqs;
		acp->prime.minorFaultCount += cb.prime.minorFaultCount;
		acp->nonprime.minorFaultCount += cb.nonprime.minorFaultCount;
		acp->prime.majorFaultCount += cb.prime.majorFaultCount;
		acp->nonprime.majorFaultCount += cb.nonprime.majorFaultCount;
		acp->prime.kcoreTime += cb.prime.kcoreTime;
		acp->nonprime.kcoreTime += cb.nonprime.kcoreTime;
		acp->prime.kvirtualTime += cb.prime.kvirtualTime;
		acp->nonprime.kvirtualTime += cb.nonprime.kvirtualTime;
		acp->prime.cpuDelayTime += cb.prime.cpuDelayTime;
		acp->nonprime.cpuDelayTime += cb.nonprime.cpuDelayTime;
		acp->prime.blkDelayTime += cb.prime.blkDelayTime;
		acp->nonprime.blkDelayTime += cb.nonprime.blkDelayTime;
		acp->prime.swpDelayTime += cb.prime.swpDelayTime;
		acp->nonprime.swpDelayTime += cb.nonprime.swpDelayTime;

		acp->ca_du += cb.ca_du;
		acp->ca_dc += cb.ca_dc;

		acp->ca_pc += cb.ca_pc;

		for(ind = 0; ind < TP_MAXDEVGRPS; ind++) {
			acp->ca_tpio[ind].cat_mounts +=
				cb.ca_tpio[ind].cat_mounts;
			acp->ca_tpio[ind].cat_reads +=
				cb.ca_tpio[ind].cat_reads;
			acp->ca_tpio[ind].cat_writes +=
				cb.ca_tpio[ind].cat_writes;
			acp->ca_tpio[ind].cat_reserv +=
				cb.ca_tpio[ind].cat_reserv;
			acp->ca_tpio[ind].cat_utime +=
				cb.ca_tpio[ind].cat_utime;
			acp->ca_tpio[ind].cat_stime +=
				cb.ca_tpio[ind].cat_stime;
		}

		acp->can_nreq += cb.can_nreq;
		acp->can_utime += cb.can_utime;
		acp->can_stime += cb.can_stime;

		acp->ca_sbu += cb.ca_sbu;
		acp->ca_fee += cb.ca_fee;
	}
}

/*
 *	Find the cacct structure which matches the consolidation options.
 *	If jid consolidation was specified, always return a new cacct 
 *	structure.  Use of the -j option is synonymous with concatenating
 *	all of the files with the "cat" command.
 */
static struct cacct *
findacd(uid_t uid, uint64_t prid, uint64_t jid, gid_t gid)
{
	static	int availind = 0;	/* 1st available entry in acmap[] */
	int ind;
	int	siz;
	int	omapsiz;

	/*
	 *	If first call allocate data struct.
	 */
	if (acmap == NULL) {
		if (db_flag >= 1) {
			fprintf(stderr, "Allocating first map structs\n");
		}
		siz = MAPSIZE * sizeof(struct map);
		mapsize = MAPSIZE;
		if ((acmap = (struct map *)malloc(siz)) == NULL) {
			acct_perr(ACCT_ABORT, errno,
				_("There was insufficient memory available when allocating '%s'."),
				"initial map");
		}

		for(ind = 0; ind < mapsize; ind++) {
			acmap[ind].uid = -1;
		}
	}

	/*
	 *	See if the consolidation keys are already in the map.
	 */
	if (!con_jid) {
	    for(ind = 0; ind < mapsize; ind++) {
		/*
		 *	If this is the end of used entries match this one.
		 */
		if (acmap[ind].uid == -1) {
			availind = ind;
			break;
		}

		/*
		 *	Look for a match. 
		 */
		if (((con_uid == 0) ||
		     ((con_uid == 1) && (acmap[ind].uid == uid))) &&
		    ((con_prid == 0) || ((con_prid == 1) && 
		     (acmap[ind].prid == prid)) || (prid == CACCT_NO_ID)) &&
		    ((con_gid == 0) || ((con_gid == 1) &&
		     (acmap[ind].gid == gid)) || (gid == CACCT_NO_ID)) ) {

			return(acmap[ind].acp); /* got it */
		}
	   }

	} else {
	   ind = availind;
	}

	if (ind < mapsize) {
		if (db_flag > 1) {
			fprintf(stderr, "findacd(): New map struct being used");
			fprintf(stderr, " jid 0x%llx, uid %d (%s), "
				"prid %lld (%s)\n", jid, uid, uid_to_name(uid),
				prid, prid_to_name(prid));
		}

		if (acmap[ind].uid == -1) {
			addmap(uid, prid, jid, gid, ind);
			if (con_jid) {
				availind++;
			}
			return(acmap[ind].acp);

		} else {
			acct_err(ACCT_ABORT,
			       _("A problem caused by a coding error %d was detected in the '%s' routine."),
				1, "findacd()");
		}
	}

	/*
	 *	If map overflowed increase size and add to map.
	 */
	if (ind == mapsize ) {
		if (db_flag > 1) {
			fprintf(stderr,"map overflowed\n");
		}
		omapsiz = mapsize;
		mapsize += MAPINC;
		siz = mapsize * sizeof(struct map);
		if((acmap = (struct map *)realloc((char *)acmap,siz)) == NULL) {
			acct_perr(ACCT_ABORT, errno,
				_("There was insufficient memory available when reallocating '%s'."),
				"map");
		}

		for(ind = omapsiz; ind < mapsize; ind++) {
			acmap[ind].uid = -1;
		}
		if (db_flag >= 1) {
			fprintf(stderr, "findacd(): expanding map table "
				"from %d to %d\n", omapsiz, mapsize);
		}
		addmap(uid, prid, jid, gid, omapsiz);
		if (con_jid) {
			availind++;
		}
		return(acmap[omapsiz].acp);

	}

	if (db_flag >= 1) {
		fprintf(stderr,"findacd(): null cacct map entry\n");
	}

	return ((struct cacct *)NULL);
}

/*
 *	Add new entry to map.
 */
static void
addmap(uid_t uid, uint64_t prid, uint64_t jid, gid_t gid, int ind)
{
	struct cacct *acp;

	acmap[ind].uid = uid;
	acmap[ind].prid = prid;
	acmap[ind].jid = jid;
	acmap[ind].gid = gid;

	/*
	 *	get memory
	 */
	if ((acp = (struct cacct *) malloc(sizeof(struct cacct))) == NULL) {
		acct_perr(ACCT_ABORT, errno,
			_("There was insufficient memory available when allocating '%s'."),
			"struct cacct");
	}

	/*
	 *	Insure zero starts.
	 */
	memset((char *)acp, '\0', sizeof(struct cacct));

	/*
	 *	Initialize ID.
	 */
	acp->ca_uid = uid;
	acp->ca_prid = prid;
	acp->ca_jid = jid;
	acp->ca_gid = gid;

	/*
	 *	Set pointer and return
	 */
	acmap[ind].acp = acp;
	return;
}

static void
wroutput() 
{
	int ind;

	if (total) {
		if (asciiout) {
			wrascii(&tcb);
		} else {
			if (create_hdr1(&tcb.ca_hdr, ACCT_CACCT) < 0) {
				acct_err(ACCT_WARN,
				       _("An error was returned from the call to the '%s' routine."),
					"header()");
			}

			if (writecacct (bofd,&tcb) != sizeof (struct cacct)) {
				acct_perr(ACCT_ABORT, errno,
					_("An error occurred during the writing of file '%s'."),
					outfile);
			}
		}
		return;
	}

	for(ind = 0; ind < mapsize; ind++) {
		if (acmap[ind].uid == -1) break;
		if (asciiout) {
			wrascii(acmap[ind].acp);
		} else {
			if (create_hdr1(&acmap[ind].acp->ca_hdr,
					ACCT_CACCT) < 0) {
				acct_err(ACCT_WARN,
				       _("An error was returned from the call to the '%s' routine."),
					"header()");
			}

			if (writecacct (bofd,acmap[ind].acp) !=
					 sizeof(struct cacct)) {
				acct_perr(ACCT_ABORT, errno,
					_("An error occurred during the writing of file '%s'."),
					outfile);
			}
		}
	}
	return;
}

char fmt[] = "%d\t%.8s\t%.8s\t%.0f\t%.0f\t%.0f\t%.0f\t%.0f\t%.0f\t%lld\t%lld\t%lld\t%lld\t%lld\t%lld\t%lld\t%lld\t%.0f\t%.0f\t%.0f\t%.0f\t%.0f\t%.0f\t";
char fmtv[] = "%d\t%.8s\t%.8s\t%e %e %e %e %e %e %lld %lld %lld %lld %lld %lld %lld %lld %e %e %e %e %e %e ";

static void
wrascii(struct cacct *tp) 
{
	int j;
	char acname[13];
	char *an;
	char *un;

	if (tp->ca_prid == CACCT_NO_ID) {
		strcpy(acname, "default");
	} else {
		if ((an = prid_to_name(tp->ca_prid)) == NULL) {
			an = "Unknown";
		}
		strncpy(acname,an,12);
		acname[12] = '\0';
	}

	if ((un = uid_to_name(tp->ca_uid)) == NULL) {
		un = "Unknown";
	}

	fprintf(ofd, verbose ? fmtv : fmt,
		tp->ca_uid, un, acname,
		tp->prime.cpuTime, tp->nonprime.cpuTime,
		tp->prime.kcoreTime, tp->nonprime.kcoreTime,
		tp->prime.kvirtualTime, tp->nonprime.kvirtualTime,
		tp->prime.charsXfr, tp->nonprime.charsXfr,
		tp->prime.logicalIoReqs, tp->nonprime.logicalIoReqs,
		tp->prime.minorFaultCount, tp->nonprime.minorFaultCount,
		tp->prime.majorFaultCount, tp->nonprime.majorFaultCount,
		tp->prime.cpuDelayTime, tp->nonprime.cpuDelayTime,
		tp->prime.blkDelayTime, tp->nonprime.blkDelayTime,
		tp->prime.swpDelayTime, tp->nonprime.swpDelayTime);

	fprintf(ofd, verbose ? "%e %lld \t%d\t%e\t%e\t" :
			"%.0f\t%lld\t%d\t%.0f\t%.0f\t",
		tp->ca_du,
		tp->ca_pc,
		tp->ca_dc,
		tp->ca_fee,
		tp->ca_sbu);

	fprintf(ofd,"\n");
}
